### ISSUE TITLE
Shipping openturns and poissoninv GPL license

### ISSUE THREAD

**Issue by mwestphal** [openturns/openturns#2557]
### What is the idea?

Reading the license of OpenTURNS and the GPL exception for its non optional component poissoninv, it looks like shipping openturns in another software would ship poinssoninv under the GPL license hence contaminating openturns and any software linked with it.

In my case, I'm interested to ship OpenTURNS in ParaView binary release.

I see two possible solutions.

1. Ask Mike Giles for an exception to the GPL for not only openturns but for ParaView as well.
2. Find a way to disable poissoninv in OpenTURNS, if that even makes sense.

Please let me know what you think before I start reaching out to Mike or putting my hand into OpenTURNS code.

Is 2. something that you may be interested in ?

**Comment by jschueller** [comment]
see COPYING.poissinv

**Comment by mwestphal** [comment]
LGPL is fine, but I don't see anything related to using open turns API in mike emails. He talks about shipping poisoning in open turns binary and with open turns source file. I don't see a specific exception related to using poisininv through open turns API.

**Comment by jschueller** [comment]
I thought that was implied

**Comment by mwestphal** [comment]
So I've been reading carefully the different GPL exception in openturns:
 1. https://github.com/openturns/openturns/blob/master/COPYING.hmat
 2. https://github.com/openturns/openturns/blob/master/COPYING.KolmogorovSmirnovDist
 3. https://github.com/openturns/openturns/blob/master/COPYING.poissinv
 
 For 1., it is clear that the exception is "in the limited from of the developement of the Open Turns software, you are allowed to used the “Hmatrix library”, recently referenced in Open Source with the GPL v3 License [instead of using the GPLv2]." which does not constitute that much of an exception, because anyone shipping OpenTurns with hmat would end up with a LGPLv3 + GPLv2 software which is not compatible. That being said, hmat is optional, so I'm not impacted.
 
 For 2. Your email initially refer to an exception the accord project got to use the code as LGPL as part of their code. Looking at accord code I found this: https://github.com/Azure/Accord-NET/blob/master/Sources/Accord.Statistics/Testing/KolmogorovSmirnovTest.cs, where there is no exception, just a straight up (re) licensing of the code as LGPL.
 So regarding the exception, Richard states (translated):
 
 ```
 You can use our program for the Kolmogorov-Smirnov under the LGPL license
 ```
 
If we assume that "you" refer either to openturns or Regis Lebrun, this does not include anyone else, and does not mention usage of Kolmogorov-Smirnov through the open turns API by other people.

For 3., Interestingly, the exchange refers to 2. and also mentions `Nobody is allowed to extract it and use it independently of OpenTURNS without your written permission, but of course anybody using OpenTURNS in a commercial product can use your code through its openTURNS interface (ie the Poisson::computeQuantile() method)`. And Mike answers `That sounds fine so I'm happy for you to use it.`

Once again, "you" refer to open turns or Regis Lebrun, and certainly not other people.

Afaics, these exception do not seem enough for people to package openturns and these GPL licensed dependencies as LGPL packages.

**Comment by jschueller** [comment]
we should clarify this, I dont know if there is a standard to formalize GPL exceptions

**Comment by mwestphal** [comment]
Here is an example: https://gitlab.onelab.info/gmsh/gmsh/blob/master/LICENSE.txt

As you can note, the exception **should be added to the upstream software**, not downstream, and the exception is targeted at specific software from specific copyright holder( eg: _ParaView from Kitware Inc_).

The exception you are suggesting that these dependencies should add are basically contournement of the code of the GPL. Indeed, if the exception says that using OpenTURNS API only does not transmit the GPL license, you could very simply expose the whole API of these dependencies in OpenTURNS completely overriding the GPL license of these dependencies.

I recommend to NOT do that, and instead ask your dependencies copyright holder (poinsininv and KolmogorovSmirnovDist) to clarify in a publicly available license outside of OpenTURNS code that "you", OpenTURNS developement team are allowed to distribute their software under another license. LGPL comes to mind.

That being said, this exception will not allow Kitware or anyone else to ship poissoninv and KolmogorovSmirnovDist so you should make them optional in your software if you want **others** to be able to ship binaries containing OpenTURNS as a LGPL package.

Let me know if it can be considered to make them optional.

**Comment by jschueller** [comment]
thats definetely ambiguous if openturns shipped outside of the OT partnership turns out to carry a different license than the one we ship
maybe its simpler to turn to boost for poissinv and kolmogorov, for hmat we probably have no replacement so we should probably warn against the potential license issue

**Comment by jschueller** [comment]
maybe @pgrt  has some insights on this

**Comment by pgrt** [comment]
Hello,

I remember these exceptions had caught my attention when first looking at the code of OpenTURNS. I have just read them again, here is how I understand the situation:

* The files of OpenTURNS incorporate the LGPL-3+ license advertisement (with some exceptions such as lib/src/Base/Algo/kissfft.h which is licensed under BSD-3-clause, lib/src/Base/Func/openturns/exprtk.hpp which is MIT, ...).

* COPYING.hmat states that it is allowed to use hmat under GPL-3 to develop OpenTURNS:
    - today hmat is under GPL-2+, so it is totally up to us to use it under GPL-3 if we wish;
    - anyway the limitation "In the limited frame of the development..." can only be restrictive.
Conclusion on the hmat matter: nowadays keeping this file COPYING.hmat is not useful, and de facto hmat contaminates OpenTURNS with the GPL. Compared with @mwestphal 's opinion: using hmat with GPL-3 makes it compatible, yet contaminant, with LGPL-3 of OpenTURNS.

* COPYING.KolmogorovSmirnovDist, in my opinion, constitutes a LGPL licensing of the code. Limitation is the LGPL version is not explicit. Compared with @mwestphal 's opinion: I tend to think the statement in the COPYING.KolmogorovSmirnovDist file grants the code under LGPL, so that anyone being granted it can also redistribute it again under LGPL, and so on... This is almost a dual licensing, "almost" because the agreement is not attached to a specific version of the code. But as long as we don't touch the embedded code copy, let's consider it is pure LGPL code.

* COPYING.poissoninv: although this is probably the intent, the upstream author does not explicitly grants a LGPL license. Like for hmat, I would tend to discard the COPYING.poissoninv file and assume the sole file from poissoninv contaminates OpenTURNS with the GPL3.

**Comment by mwestphal** [comment]
>  This is almost a dual licensing, "almost" because the agreement is not attached to a specific version of the code. But as long as we don't touch the embedded code copy, let's consider it is pure LGPL code.

TBH I agree with you.

However, exception should be granted upstream, for specific usage, od the code should clearly relicensed or dual licensed in the upstream. Although the intent is clear in the mail exchange, that will not be enough for Kitware and ParaView to ship OpenTURNS with our binary while KolmogorovSmirnovDist is needed.

Let alone poissoninv.

Lets hope #2564 can move forward and be integrated soon :)

**Comment by pgrt** [comment]
Plainly agreeing :)

**Comment by pgrt** [comment]
Maybe for hmat we should try to either have it relicensed under LGPL or keep GPL with something like the OpenSSL exception of edbrowse, see https://sources.debian.org/src/edbrowse/3.7.7-5/debian/copyright/?hl=11#L11 lines 11 to 24 for the version of edbrowse currently in Debian.

**Comment by jschueller** [comment]
- [x] poissinv (#2564)
- [ ] KolmogorovSmirnovDist
- [ ] hmat

**Comment by jschueller** [comment]
for KolmogorovSmirnovDist
scipy has an implementation of the simard method it is possible to translate to c++:
https://github.com/scipy/scipy/blob/main/scipy/stats/_ksstats.py

however it also relies on some C code from the cephes library from which they obtained permission from the author
https://github.com/scipy/scipy/blob/main/scipy/special/cephes/kolmogorov.c
and that's unclear for me whether we can redistribute in OT and transitively in paraview, any clue @pgrt ?

**Comment by pgrt** [comment]
Hi @jschueller ,

I might be overlooking this, but I feel the cephes code in scipy is licensed under BSD-3-Clause, so we should be able to use it in OT while keeping its LGPL-3?

**Comment by jschueller** [comment]
yes, but its a "with permission from the author", so we should ask the same permission or is it enough ?
https://github.com/scipy/scipy/blob/main/LICENSES_bundled.txt#L97

**Comment by pgrt** [comment]
Thanks, I had missed that.

When the author gave his permission, he licensed the exact code that was included in scipy under BSD-3-Clause. So if we pick this code in the source of scipy then it is licensed under BSD-3-Clause, thus free and compatible with a bunch of free licenses.

However, if the version of Cephes in scipy has changed since the approval of the author, then we cannot be sure he also properly relicensed the newer version and we should not use it a priori.

In short:
- if we intend to use a version of Cephes that is newer than the one that got author's approval, we need to ask him again for this version;
- if we rely on the portion of the exact version of the Cephes code that was relicensed using author's approval, then it is pure BSD-3-Clause and we can use it!

**Comment by jschueller** [comment]
thanks for the feedback, I sent a mail but did not receive an answer yet

**Comment by mwestphal** [comment]
Thanks! You could specify in the main readme that all other licenses are permissive or something similar in order to help people like me when they check if they can use a software or not :)

**Comment by jschueller** [comment]
yes, I updated the LICENSE file with the licenses of each bundled third-party codes
